nessie/src/test/java/org/apache/iceberg/nessie/TestNessieTable.java

```diff
@@ -0,0 +1,348 @@
+    try (FileAppender<GenericData.Record> writer = Avro.write(Files.localOutput(location2))
+        .schema(schema)
+        .named("test")
+        .build()) {
+      for (GenericData.Record rec : records) {
+        writer.add(rec);
+      }
+    }
+
+    DataFile file1 = DataFiles.builder(table.spec())
+        .withRecordCount(3)
+        .withPath(location1)
+        .withFileSizeInBytes(Files.localInput(location2).getLength())
+        .build();
+
+    DataFile file2 = DataFiles.builder(table.spec())
+        .withRecordCount(3)
+        .withPath(location2)
+        .withFileSizeInBytes(Files.localInput(location1).getLength())
+        .build();
+
+    // add both data files
+    table.newAppend().appendFile(file1).appendFile(file2).commit();
+
+    // delete file2
+    table.newDelete().deleteFile(file2.path()).commit();
+
+    String manifestListLocation =
+        table.currentSnapshot().manifestListLocation().replace("file:", "");
+
+    List<ManifestFile> manifests = table.currentSnapshot().allManifests();
+
+    Assert.assertTrue(catalog.dropTable(TABLE_IDENTIFIER));
+    Assert.assertFalse(catalog.tableExists(TABLE_IDENTIFIER));
+
+    Assert.assertTrue(new File(location1).exists());
+    Assert.assertTrue(new File(location2).exists());
+    Assert.assertTrue(new File(manifestListLocation).exists());
+    for (ManifestFile manifest : manifests) {
+      Assert.assertTrue(new File(manifest.path().replace("file:", "")).exists());
+    }
+    Assert.assertTrue(new File(
+        ((HasTableOperations) table).operations()
+            .current()
+            .metadataFileLocation()
+            .replace("file:", ""))
+        .exists());
+  }
+
+  @Test
+  public void testExistingTableUpdate() {
+    Table icebergTable = catalog.loadTable(TABLE_IDENTIFIER);
+    // add a column
+    icebergTable.updateSchema().addColumn("data", Types.LongType.get()).commit();
+
+    icebergTable = catalog.loadTable(TABLE_IDENTIFIER);
+
+    // Only 2 snapshotFile Should exist and no manifests should exist
+    Assert.assertEquals(2, metadataVersionFiles(TABLE_NAME).size());
+    Assert.assertEquals(0, manifestFiles(TABLE_NAME).size());
+    Assert.assertEquals(altered.asStruct(), icebergTable.schema().asStruct());
+
+  }
+
+  @Test
+  public void testFailure() throws NessieNotFoundException, NessieConflictException {
+    Table icebergTable = catalog.loadTable(TABLE_IDENTIFIER);
+    Branch branch = (Branch) client.getTreeApi().getReferenceByName(BRANCH);
+
+    IcebergTable table = client.getContentsApi().getContents(KEY, BRANCH).unwrap(IcebergTable.class).get();
+
+    client.getContentsApi().setContents(KEY, branch.getName(), branch.getHash(), "",
+        IcebergTable.of("dummytable.metadata.json"));
+
+    AssertHelpers.assertThrows("Update schema fails with conflict exception, ref not up to date",
+        CommitFailedException.class,
+        () -> icebergTable.updateSchema().addColumn("data", Types.LongType.get()).commit());
```

Comment:
Do you have a test for concurrent writes to the same table from multiple catalogs? That would be good with two cases: two tables that share the same ref (loaded by the same catalog) and two tables that were loaded by different catalogs and have separate refs.

Reply:
`TestCatalogBranch` does this for the positive case (ie concurrent changes that succeed) and internal nessie for teh negative case. I have added a negative case to `TestCatalogBranch` also. (and renamed the class)

Reply:
Cleaned this up a bit more. We do both tests you suggested, thought it may not be clear its happening. Let me know if the extra comments aren't enough.